### DISCUSSION TITLE
Difference with VSCodeVim w/ Neovim Integration

### DISCUSSION THREAD

**Discussion by Malix-Labs** [vscode-neovim/vscode-neovim#1789] [Q&A] [answered]
As https://github.com/VSCodeVim/Vim/issues/1735 exists, how does VSCodeVim with Neovim Integration and vscode-neovim differs?
VSCodeVim does work web-only, but I don't know more

**Comment by justinmk** [comment, marked as answer]
As [mentioned in the VSCodeVim readme](https://github.com/VSCodeVim/Vim?tab=readme-ov-file#neovim-integration), it only uses Nvim for Ex commands (like `:sub`, `:global`, etc.).

vscode-neovim uses Nvim for everything (except insert-mode, which is owned by VSCode).

**Reply by Malix-Labs** [reply]
Ohkay

What would be the best way to make a setup that would both use neovim when possible but that would work with emulation on VSCode Web ?

**Reply by justinmk** [reply]
~~vscode.dev~~ github codespaces has a mode with "compute" where normal `nvim` can run and works just fine with vscode-neovim.

If you mean "vscode web" in a browser-only context (no "compute"), it likely makes sense to explore compiling `nvim` to WASM so it can run in a browser-only context.

If you are asking how to enable vscode extensions "conditionally", that's more like a feature request for vscode. I don't know if that's supported by vscode currently. Although you can install vscode extensions to the "workspace" without installing them locally.

**Reply by Malix-Labs** [reply]
Thanks for those pov

Yeah, I meant conditionally

**Reply by theol0403** [reply]
He just means using a web-based vscode environment which has an integrated terminal and is running on an underlying linux container. Github codespaces is such an environment, but I do not know if vscode.dev has anything other than browser-only.

**Reply by theol0403** [reply]
Another such environment is using code-server on your own server.

**Reply by Malix-Labs** [reply]
@theol0403 do you think it would be possible for vscode-neovim to detect when the client is vscode.dev (/ github.dev) and to use a [WASM version of neovim](https://github.com/vscode-neovim/vscode-neovim/discussions/1789#discussioncomment-8429683) in that case ?

If yes, should I create an issue (feature request case) about it ?

**Reply by justinmk** [reply]
sure, create an issue. But there is no WASM build of Nvim, yet. Help welcome...

**Reply by Malix-Labs** [reply]
> there is no WASM build of Nvim

Ah shit, do you know if it would even be possible ?

**Reply by justinmk** [reply]
there is a prototype here: https://nvim.nry.app/ by @naruaway 

p.s Naru: did you publish a repo that shows how you did this? :)

**Reply by Malix-Labs** [reply]
_Btw, is VSCodeVim using a vim dependency, or is it just replicating its feature in JavaScript ?_

**Reply by justinmk** [reply]
See [previous answer](https://github.com/vscode-neovim/vscode-neovim/discussions/1789#discussioncomment-8429471).

If VSCodeVim used vim or nvim as a backend, then vscode-neovim wouldn't be needed.

**Reply by Malix-Labs** [reply]
The previous answer stated that VSCodeVim used Nvim only for Ex commands, not that it is using either Vim or a JavaScript vim-motion implementation in all other cases

My guess is that VSCodeVim is using a JavaScript vim-motion implementation in all other cases, and can be set to use nvim for Ex commands

**Reply by justinmk** [reply]
VSCodeVim reimplements everything from scratch. That's why it uses Nvim for Ex commands. If it had a Vim backend it wouldn't need a (optional) Nvim backend.
